### PR TITLE
Add "Enter Nickname" and "Waiting Room" designs for the join room page.

### PR DESCRIPTION
**Pull request by zpChris**
### List of Changes:
- Add functionality to the join room page to dynamically display and update the "Enter Nickname" and "Waiting Room" pages as necessary (as well as switch back and forth between the two).
- Create the design for the "Enter Nickname" page.
  - Add new text and button input elements.
  - Add input button styling for when the input button is disabled.
  - Allow for validity-checking on the user-entered nickname. When the user's current nickname is invalid and they are on the input text element, it shows these in the UI (see the "Screencast and Images" section below). I chose the following parameters for nickname validity:
    1. The nickname must be non-empty.
    2. The nickname must have no spaces.
    3. The nickname length must be less than 16 characters.
  - Upon entering a nickname, the user either presses the `Enter` key or clicks the "Enter" button; this connects them to the socket and sends them to the "Waiting Room" page.
- Add the "Waiting Room" page (which will just hold the current user's nickname").

### Notes:
- The branch name is different than the actual contents, sorry about that -- I had initially anticipated to do more socket work in this PR. 
- This PR removed the calls to the `Socket.ts` functions as they weren't relevant yet; I will be using these functions in the next PR.
- I used a `switch` block for the `pageState`, even though there are only two stages currently (and could easily use a `boolean` `if / else`, because eventually the Join Room page will have another `pageState` for the "Enter Room ID" page, so I wanted to design the code keeping future development in mind.
- The "waiting room" page will eventually list all of the users that are currently on that page.
- In the future, we could make it possible to update one's nickname, but for the MVP I wanted to keep it to the basic features.
- The join room page may eventually use the `useLocation` property from `'react-router-dom'` so the nickname is maintained even after you leave the page; however, I wanted to keep it simple for now, as this PR is just going to be used to complete the socket and user connections work (see [this Trello task](https://trello.com/c/lHMJjRQ9/44-determine-everybody-that-is-on-a-url)).

### Screencast and Images:
[Screencast of the Join Room Page](https://www.loom.com/share/6638a7ae004d44f9901990741339ce53)

Enter Nickname Page with no Input (and nickname validation text shown)
Enter Nickname Page with no Input

Enter Nickname Page with "ChrisElliott" Input
Enter Nickname Page with Input

Waiting Room Page with "ChrisElliott" Nickname
Waiting Room Page

### CODE REVIEW

File: frontend/src/views/Join.tsx
Comment (zpChris) [inline frontend/src/views/Join.tsx:33]: I'm open to changing the design of this portion of the file if it is too complex, let me know your thoughts (I'm also happy to add more comments)!

**Comment by alankbi** [comment]
Love the design, I'm going to hold off on reviewing until SocketSubscribe is merged into master. I also think it might be good to discuss flow a little bit more (maybe at the next sprint review/planning) which might end up affecting this PR a bit.

File: frontend/src/views/Join.tsx
Comment (alankbi) [reply frontend/src/views/Join.tsx:33]: I think maybe the waiting room component might eventually be moved to another page, but for the sake of this PR and the waiting room/join room page designs, I think this looks great

File: frontend/src/views/Join.tsx
Comment (alankbi) [inline frontend/src/views/Join.tsx:52]: Note: the logic here might change in the next few PRs right? I'm assuming we'd want this to trigger a REST call to join the room and then after getting to the waiting room then connecting to the socket. This may be looking too far ahead though

**Review by alankbi — APPROVED** [review]
Looks great, only one nit and a couple of discussion comments but otherwise approved!

The last two notes I have: 

1. When refreshing the page, it resets you back to the nickname page. I'm assuming that's fine for now, but once we get to hooking up the actual functionality that we might have some protection against that.

2. When we merge this PR, do we have to open another PR from SocketSubscribe into master, or will it go straight into master (rn it still shows as merging into SocketSubscribe even though the other PR has been merged)?

**Review by helenli8181 — APPROVED** [review]
This all looks great to me! :) I noticed that you used a Function with useState(), while in one of my PRs I had a Class with setState(). Later could we discuss which of these options should be used? I may end up changing mine to a function if that's better.

File: frontend/src/components/config/Theme.tsx
Comment (zpChris) [reply frontend/src/components/config/Theme.tsx:11]: Agreed; I had actually tried this on past PRs with [Coolors](https://coolors.co/u/chris_elliott) as well, but ended up not incorporating it. I thought the 10-color colorful palette could be used to create the different users' colors. I agree that we should use this in the future!

File: frontend/src/views/Join.tsx
Comment (zpChris) [reply frontend/src/views/Join.tsx:52]: Yes, that's right -- since I didn't have those components ready yet, I simply added an instant connection to a predefined endpoint upon pressing "Enter".

**Comment by zpChris** [comment]
> This all looks great to me! :) I noticed that you used a Function with useState(), while in one of my PRs I had a Class with setState(). Later could we discuss which of these options should be used? I may end up changing mine to a function if that's better.

Yes, I agree; we've talked about this at [this discussion](https://github.com/rocketden/main/pull/22#discussion_r484538483). I think we should stick with React Hooks (which means  functions with `useState()` -- see that discussion for details and reasons.

File: frontend/src/components/core/Input.tsx
Comment (zpChris) [reply frontend/src/components/core/Input.tsx]: Thanks for the note -- done!

**Comment by zpChris** [comment]
> Looks great, only one nit and a couple of discussion comments but otherwise approved!
> 
> The last two notes I have:
> 
> 1. When refreshing the page, it resets you back to the nickname page. I'm assuming that's fine for now, but once we get to hooking up the actual functionality that we might have some protection against that.
> 2. When we merge this PR, do we have to open another PR from SocketSubscribe into master, or will it go straight into master (rn it still shows as merging into SocketSubscribe even though the other PR has been merged)?

1. I agree completely; we can use the `useLocation` ability to save the current state of the page. There will also be some decisions we have to make on how to reconnect the sockets in that case -- some code design will need to be done there.
2. That's because I forgot to delete the branch after merging the PR; sorry about that, just fixed! You can now see the update that "Base automatically changed from `SocketSubscribe` to `master`".

**Comment by zpChris** [comment]
[Create join room + Enter nickname pages](https://trello.com/c/yhty0iVw/49-create-join-room-enter-nickname-pages)

**Pull request merged**
